sourcerer-core/src/main/java/org/elder/sourcerer2/ExpectedVersion.kt

```diff
@@ -0,0 +1,127 @@
+            return if (baseVersion != null && newVersion != null) {
+                if (baseVersion.ordinal() > newVersion.ordinal()) {
+                    mergeOrdered(newVersion, baseVersion)
+                } else {
+                    mergeOrdered(baseVersion, newVersion)
+                }
+            } else baseVersion ?: (newVersion ?: ExpectedVersion.any())
+        }
+
+        /**
+         * Assumes the expected versions are given in ordinal order.
+         */
+        private fun mergeOrdered(
+                baseVersion: ExpectedVersion,
+                newVersion: ExpectedVersion
+        ): ExpectedVersion {
+            return when (baseVersion) {
+                ExpectedVersion.Any ->
+                    // Any is always compatible with other expected version
+                    newVersion
+                ExpectedVersion.AnyExisting -> when (newVersion) {
+                    ExpectedVersion.AnyExisting,
+                    ExpectedVersion.Any -> ExpectedVersion.anyExisting()
+                    ExpectedVersion.NotCreated -> throw ConflictingExpectedVersionsException(
+                            "Cannot merge 'any existing' with 'not created'",
+                            baseVersion, newVersion)
+                    is ExpectedVersion.Exactly -> newVersion
+                }
+                is ExpectedVersion.Exactly -> when (newVersion) {
+                    is ExpectedVersion.Exactly -> {
+                        if (baseVersion.streamVersion != newVersion.streamVersion) {
+                            throw ConflictingExpectedVersionsException(
+                                    "Cannot merge 'exactly "
+                                            + baseVersion.streamVersion
+                                            + "' with 'exactly "
+                                            + newVersion.streamVersion
+                                            + "'",
+                                    baseVersion, newVersion)
+                        }
+                        baseVersion
+                    }
+                    ExpectedVersion.NotCreated ->
+                        throw ConflictingExpectedVersionsException(
+                                "Cannot merge 'exactly "
+                                        + baseVersion.streamVersion
+                                        + "' with 'not created'",
+                                baseVersion, newVersion)
+                    else ->
+                        throw RuntimeException(
+                                "Internal error, unexpected version combination merging "
+                                        + baseVersion
+                                        + " with " + newVersion)
+
+                }
+                ExpectedVersion.NotCreated -> when (newVersion) {
+                    ExpectedVersion.NotCreated -> baseVersion
+                    else ->
+                        throw RuntimeException(
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Caught exception is too generic. Prefer catching specific Exceptions to handle specific error cases. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.elder.org/coding_rules#rule_key=detekt-kotlin%3AThrowRuntimeException)